cps/kni.c

```diff
@@ -850,6 +850,273 @@ rd_send_err(const struct nlmsghdr *req, int err)
+	rm->rtm_src_len = 0;
+	rm->rtm_tos = 0;
+	rm->rtm_table = RT_TABLE_MAIN;
+	rm->rtm_scope = RT_SCOPE_UNIVERSE;
+	rm->rtm_type = RTN_UNICAST;
+	rm->rtm_flags = 0;
+
+	switch (fib->action) {
+	case GK_FWD_GRANTOR:
+		mnl_attr_put_u32(reply, RTA_OIF, cps_conf->back_kni_index);
+		rm->rtm_protocol = RTPROT_STATIC;
+		*gw_addr = &fib->u.grantor.eth_cache->ip_addr;
+		break;
+	case GK_FWD_GATEWAY_FRONT_NET:
+		mnl_attr_put_u32(reply, RTA_OIF, cps_conf->front_kni_index);
+		rm->rtm_protocol = fib->u.gateway.rt_proto;
+		*gw_addr = &fib->u.gateway.eth_cache->ip_addr;
+		break;
+	case GK_FWD_GATEWAY_BACK_NET:
+		mnl_attr_put_u32(reply, RTA_OIF, cps_conf->back_kni_index);
+		rm->rtm_protocol = fib->u.gateway.rt_proto;
+		*gw_addr = &fib->u.gateway.eth_cache->ip_addr;
+		break;
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		mnl_attr_put_u32(reply, RTA_OIF, cps_conf->front_kni_index);
+		rm->rtm_protocol = RTPROT_UNSPEC;
+		*gw_addr = NULL;
+		break;
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		mnl_attr_put_u32(reply, RTA_OIF, cps_conf->back_kni_index);
+		rm->rtm_protocol = RTPROT_UNSPEC;
+		*gw_addr = NULL;
+		break;
+	default:
+		rte_panic("Invalid FIB action (%u) in FIB while being processed by CPS block in %s\n",
+			fib->action, __func__);
+		return;
+	}
+}
+
+static int
+rd_send_batch(struct cps_config *cps_conf, struct mnl_nlmsg_batch *batch,
+	const char *daemon, uint32_t seq, uint32_t pid, int done)
+{
+	/* Address of routing daemon. */
+	struct sockaddr_nl rd_sa;
+	size_t len = mnl_nlmsg_batch_size(batch);
+
+	memset(&rd_sa, 0, sizeof(rd_sa));
+	rd_sa.nl_family = AF_NETLINK;
+	rd_sa.nl_pid = pid;
+
+	if (done) {
+		struct nlmsghdr *done =
+			mnl_nlmsg_put_header(mnl_nlmsg_batch_current(batch));
+		done->nlmsg_type = NLMSG_DONE;
+		done->nlmsg_flags = NLM_F_MULTI;
+		done->nlmsg_seq = seq;
+		done->nlmsg_pid = cps_conf->nl_pid;
+		len += sizeof(*done);
+	}
+
+	if (sendto(mnl_socket_get_fd(cps_conf->rd_nl),
+			mnl_nlmsg_batch_head(batch), len, 0,
+			(struct sockaddr *)&rd_sa, sizeof(rd_sa)) < 0) {
+		int saved_errno = errno;
+		CPS_LOG(ERR,
+			"sendto: cannot dump route batch to %s daemon (pid=%u seq=%u): %s\n",
+			daemon, pid, seq, strerror(errno));
+		return -saved_errno;
+	}
+
+	return 0;
+}
+
+static int
+rd_getroute_ipv4_locked(struct cps_config *cps_conf, struct gk_lpm *ltbl,
+	struct mnl_nlmsg_batch *batch, const struct nlmsghdr *req, int family)
+{
+	struct rte_lpm_iterator_state state;
+	const struct rte_lpm_rule *re4;
+	int index;
+
+	int ret = rte_lpm_iterator_state_init(ltbl->lpm, 0, 0, &state);
+	if (ret < 0) {
+		CPS_LOG(ERR, "Failed to initialize the IPv4 LPM rule iterator state in %s\n",
+			__func__);
+		return ret;
+	}
+
+	index = rte_lpm_rule_iterate(&state, &re4);
+	while (index >= 0) {
+		struct gk_fib *fib = &ltbl->fib_tbl[re4->next_hop];
+		struct ipaddr *gw_addr;
+		struct nlmsghdr *reply;
+
+		if (fib->action != GK_FWD_GATEWAY_FRONT_NET &&
+				fib->action != GK_FWD_GATEWAY_BACK_NET &&
+				fib->action != GK_FWD_GRANTOR) {
```

Comment:
We no longer need this `if` statement.

This comment also applies to the IPv6 case.